### ISSUE TITLE
expression.py hard coded record ids in sql lead to low performance

### ISSUE THREAD

**Issue by netice** [odoo/odoo#12801]
Impacted versions:
8.0 and so on

Steps to reproduce:
create a many2many field in product.product such as:
`'company_ids':fields.many2many('res.company','res_company_product_rel','partner_id','company_ids',string=u'multi company''),`
and add a ir.rule for product.product with  domain such as
`['|',('company_id','=',False),('company_ids','in',[c.id for c in user.company_ids])]`
and import 10 thousand product.
then login with normal user(not super) and open the product.produt list view
it taks several minus to open

Current behavior:
when I search product.product
all product id have been hard coded in the sql 
lead to low performance.

Expected behavior:
not hard code produt id

Video/Screenshot link (optional):
in the sql below 10 thousand product id have been hard code in the position %s
AND ("product_product__product_tmpl_id"."company_id" IS NULL   OR  ("product_product"."id" in (%s)))

**Comment by gdgellatly** [comment]
@odony I find this one interesting.  I do wonder whether it is worth a special case on these company type access rules to have that query resolve to product_product__product_tmpl_id.company_id IN (ids).  I thought that's what auto_join would do but maybe it hasn't.  Would a chained domain rule like (product_tmpl_id.company_id, in, user_company_ids) resolve?

**Comment by dreispt** [comment]
I was about to open an Issue to report/discuss a similar issue:
In a database with many thousands of Project Tasks, and their many Followers, I have users that wait for a few minutes to load the Tasks Kanban.

This is caused by record rules with `('message_follower_ids', 'in', [user.partner_id.id])`, generating in the raw SQL a condition `id in (x, y, x, ...)` that can use thousands of ids. I'm sure that this is `in`clause is very inefficient, because without it a query taking many seconds becomes almost instantaneous.

BTW, for this particular case, I traced the  `in` condition to the [mail_thread._search_follower_partners](https://github.com/odoo/odoo/blob/9.0/addons/mail/models/mail_thread.py#L131) method.

I wonder: why specific method was needed for `mail_thread` followers?
Could we skip the method and use `auto_join`?
Or maybe somehow replace the IDs list with a subselect, that is probably handled more efficiently by Postgres?

With the current implementation, record level access security based of Followers is not scalable at all.
If not fix is possible I will be forced to bypass it completely, since the user experience is not acceptable.

Would really like to hear your comments on this @odony

**Comment by netice** [comment]
i have fixed it  myself

**Comment by Cedric-Pigeon** [comment]
Hello,

I had the same problem with company_ids field on product. It was really difficult to use some features.
For example using the transfert wizard on a picking with more than 200 rows took more than 8 minutes !
I fixed it by myself to force Odoo using a join in the apply rule query.
Other many2many like that causes some performance issue on my customer platform.

A sustainable solution like using auto_join will be really appreciated.

Thanks

**Comment by Yenthe666** [comment]
@odony and @mart-e could you guys shed some light from the Odoo side on this one please?

**Comment by dreispt** [comment]
Can't we have a `IN (SELECT id FROM ...)` instead of a `in (<ids>...)` ?

**Comment by netice** [comment]
I have fixed it as IN (SELECT id FROM ...)
you can get the patch from my blog
http://www.designdock.cn/blog/2/post/oe-domain-bug-5











在2016年07月25 22时01分，"Daniel Reis"<notifications@github.com>写道：

Can't we have a IN (SELECT id FROM ...) instead of a in (<ids>...) ?

**Comment by JKE-be** [comment]
@dreispt 
The search method allow the operator inselect.
You can easily change the inselect for thread.res_id but not for res_partner, because you need to convert the domain to sql.
You will avoid the `'res' in ([res_id,res_id,res_id,...])`, but that will use anyway the `'res' in (select id from res_partner where id in [id,id,id,...])`

```python
dom = [
    ('res_model', '=', self._name),
    ('partner_id', operator, operand)
]

cr, uid, ctx = self.env.cr, self.env.uid, self.env.context
e = expression(cr, uid, dom, self.env['mail.followers'], ctx)
where_clause, where_params = e.to_sql()
req = ('SELECT res_id FROM mail_followers WHERE %s' % where_clause, where_params)
return [('id', 'inselect', req)]
```

**Comment by rco-odoo** [comment]
Fixed with 3c2065c89da757406f080d9252e2c3019c66e8c7.
